torch/distributed/_fsdp/fully_sharded_data_parallel.py

```diff
@@ -0,0 +1,803 @@
+        """
+        if memo is None:
+            memo = set()
+        for module in self.modules():
+            if module is not self and isinstance(module, FullyShardedDataParallel):
+                # Allow any child FSDP instances to handle their own buffers.
+                module._cast_buffers(device=device, memo=memo)
+            elif module not in memo:
+                memo.add(module)
+                for name, buf in module.named_buffers(recurse=False):
+                    if buf is None:
+                        continue
+                    buf = buf.to(device=device or self.compute_device)
+                    setattr(module, name, buf)
+
+    @torch.no_grad()
+    def _shard_parameters_(self) -> None:
+        """
+        At initialization we wrap a module with full parameters and shard the
+        parameters in-place. Sharding is implemented by viewing each parameter
+        as a 1D Tensor and retaining only a single slice, where the slice size
+        is determined by the number of data parallel workers.
+        After this initial sharding is complete, the user can initialize a
+        ``torch.optim.Optimizer`` in the usual way, i.e.::
+        .. code-block:: python
+            optim = torch.optim.Adam(sharded_module.parameters(), lr=0.0001)
+        The optimizer will see only a single slice of parameters and will thus
+        allocate less memory for optimizer state, avoiding redundancy across
+        data parallel workers.
+        """
+        self.numel_padded_per_param = []
+        for p in self.params:
+            assert not hasattr(p, "_is_sharded")
+            assert p.is_floating_point()
+
+            # Sharding is done only when world_size is larger than 1.
+            p._is_sharded = self.world_size > 1  # type: ignore[attr-defined]
+            p._orig_size = p.size()  # type: ignore[attr-defined]
+
+            if not p._is_sharded:  # type: ignore[attr-defined]
+                self.numel_padded_per_param.append(0)
+                continue
+
+            # Replace p.data with the relevant shard.
+            orig_data = p.data
+            local_shard, num_padded = self._get_shard(p)
+            with torch.no_grad():
+                p.set_(local_shard)  # type: ignore[call-overload]
+            self.numel_padded_per_param.append(num_padded)
+            _free_storage(orig_data)
+        assert len(self.numel_padded_per_param) == len(self.params)
+
+    def _get_shard(self, tensor: torch.Tensor) -> Tuple[torch.Tensor, int]:
+        """Return the local shard of a full tensor."""
+        # Shard using torch.chunk to match all-gather/reduce-scatter.
+        chunks = list(torch.flatten(tensor).chunk(self.world_size))
+        if len(chunks) < (self.rank + 1):
+            chunk = chunks[0].new_empty(0)
+        else:
+            chunk = chunks[self.rank]
+        # Determine number of padding elements.
+        num_to_pad = chunks[0].numel() - chunk.numel()
```

Comment:
This computation is incorrect if `chunk is chunks[0]` and it is not full size right?

Reply:
chunks[0] is always full size and not padded, right?

Reply:
yes, it always filled the first chunks into full size first. Tried these examples: 

import torch
a = torch.randn(2)
a_2_5 = a.chunk(5)
print(f"size 2 chunked by 5: {a_2_5}")
a = torch.randn(6)
a_6_5 = a.chunk(5)
print(f"size 6 chunked by 5: {a_6_5}")

size 2 chunked by 5: (tensor([0.1430]), tensor([1.2853]))
size 6 chunked by 5: (tensor([ 0.7028, -0.3949]), tensor([-0.9336,  0.3286]), tensor([-0.2723,  1.2107]))